robots/tank_netsink.h

```diff
@@ -31,16 +31,19 @@ class TankNetSink : public Tank
         }
 
         // send steering command
-        m_Node->getSocket()->send("TNK " + std::to_string(left) + " " +
-                                  std::to_string(right) + "\n");
+        {
+            std::lock_guard<Net::Node> guard(m_Node);
```

Comment:
I think leaving the locking of this up to users of ``Net::Node`` is a bit error-prone. It would be nicer if you made ``Node::getSocket()`` private and added wrappers with ``std::lock_guard`` around the ``Net::Socket`` functions that users of ``Net::Node`` actually should be using (i.e. send and receive) to ``Net::Node``

Reply:
I haven't made this particularly apparent, but I've changed the code so that we only do locking for sending now, as reading always takes place on the same thread and some of the handlers for incoming commands need to send stuff.

I did have it the way you suggested before, but the problem is that when we're sending and receiving frames we need to do two reads/writes in quick succession (one for the ASCII command, one for the raw JPEG bytes that come after) and sometimes we could end up with another read or write in the middle which would screw things up. I did think of a more C++-y way of forcing users to lock it: return a little object called ``LockedSocket`` or something, which just provides a ``get()`` method to get at the ``Socket*`` and locks and unlocks the ``Node``'s mutex on construction/destruction. What do you think?

Reply:
I suppose the downside to this approach is that we'll probably end up with a bunch of unnecessary locking and unlocking of the mutex

Reply:
Ahh yeah of course - in that case I think that what we have now is fine. Although it would be kinda cool if you hid the socket like I said previous and added a method which returned one of these as a kinda socket writing transaction object:

```c++
class SocketWriter
{
public:
    SocketWriter(Socket &socket, std::mutex &mutex) : m_Socket(socket), m_Guard(mutex)
    {
    }
    SocketWriter(const SocketWriter&) = delete;

    void write(...){ m_Socket.write(...); }

private:
    Socket &m_Socket;
    std::lock_guard<std::mutex> m_Guard;
};
```